### PR TITLE
Add shell completion command to cli

### PR DESCRIPTION
**Pull request by siggy**
Cobra supports bash and zsh completion code generation, but our cli
project was not leveraging it.

Add a 'completion' command to the conduit cli, which outputs bash and
zsh completion code.

Signed-off-by: Andrew Seigner <andrew@sig.gy>

Note that the completion code can be further customized and embellished using Cobra annotations, but this PR gets us started.

Example usage:

```bash
$ source <(conduit completion bash)

$ conduit [tab][tab]
completion  dashboard   get         inject      install     stat        tap         version

$ conduit completion [tab][tab]
bash  zsh

$ conduit inject -[tab][tab]
--api-port=             --conduit-version=      --image-pull-policy=    --init-image=           --proxy-image=          --skip-inbound-ports=   -n
--conduit-namespace=    --control-port=         --inbound-port=         --outbound-port=        --proxy-uid=            --skip-outbound-ports=  -v

$ conduit install -[tab][tab]
--conduit-namespace=    --controller-replicas=  --image-pull-policy=    --prometheus-replicas=  --registry=             --version=              --web-replicas=         -n                      -r                      -v
```

### CODE REVIEW

File: cli/cmd/completion_test.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+		if !strings.Contains(zsh, "#compdef conduit") {
+			t.Fatalf("Unexpected zsh outpout: %+v", zsh)
+		}
+	})
+
+	t.Run("Fails with invalid shell type", func(t *testing.T) {
+		out, err := getCompletion("foo")
+		if err == nil {
+			t.Fatalf("Unexpected success for invalid shell type: %v", out)
```
Comment (pcalcado) [inline cli/cmd/completion_test.go]: No `+`? I'm thinking we should make it the standard

File: cli/cmd/completion.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	"log"
+
+	"github.com/spf13/cobra"
+)
+
+var completionCmd = &cobra.Command{
+	Use:       "completion [bash|zsh]",
+	Short:     "Shell completion",
+	Long:      "Output completion code for the specified shell (bash or zsh).",
+	Args:      cobra.ExactArgs(1),
```
Comment (pcalcado) [inline cli/cmd/completion.go]: 😍 I didn't know about this. We should change other commands to use its variations.

File: cli/cmd/completion.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	"fmt"
+	"log"
+
+	"github.com/spf13/cobra"
+)
+
+var completionCmd = &cobra.Command{
+	Use:       "completion [bash|zsh]",
+	Short:     "Shell completion",
+	Long:      "Output completion code for the specified shell (bash or zsh).",
```
Comment (pcalcado) [inline cli/cmd/completion.go]: What about adding a  `source <(conduit completion bash)` copy-and-pastable line?

**Comment by pcalcado** [comment]
Awesome!

I am not sure if I'm doing something wrong, but I couldn't get it to work on my shell(iTerm 2 FWIW) with this branch.

On bash:
```
bash-3.2$ bash
bash-3.2$  source <(conduit completion bash)
bash-3.2$ conduit [tab] [tab]
.CFUserTextEncoding          .config/                     .gitmessage                  .local/                      .rustup/                     .wget-hsts                   .zshrc                       Movies/                      tmp/
[...]
bash-3.2$ conduit completion bash | head && echo $?
# bash completion for conduit                              -*- shell-script -*-

__debug()
{
    if [[ -n ${BASH_COMP_DEBUG_FILE} ]]; then
        echo "$*" >> "${BASH_COMP_DEBUG_FILE}"
    fi
}

# Homebrew on Macs have version 1.3 of bash-completion which doesn't include
0
bash-3.2$ bash -version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin17)
Copyright (C) 2007 Free Software Foundation, Inc.
```

And with oh-my-zsh:

```
pcalcado@arbeitsziege.home:/Users/pcalcado  $ source <(conduit completion zsh)
_arguments:comparguments:319: can only be called from completion function
_arguments:comparguments:319: can only be called from completion function
pcalcado@arbeitsziege.home:/Users/pcalcado  $ zsh --version
zsh 5.3 (x86_64-apple-darwin17.0)
```

Am I doing something funky?

**Comment by siggy** [comment]
@pcalcado  thanks for testing, looks like `source <(conduit completion bash)` won't work on bash 3.2 (most macs). instead run  `source /dev/stdin <<< "$(conduit completion bash)"`.

reference:
- http://lists.gnu.org/archive/html/bug-bash/2006-01/msg00018.html

File: cli/cmd/completion.go (hunk 2 above)
Comment (siggy) [reply cli/cmd/completion.go]: yeah it's a nice change. also available is `Args: cobra.OnlyValidArgs()`, which works in conjunction with `ValidArgs: []string{"bash", "zsh"}`.

but what I really want is `Args: cobra.ExactArgs(1) && cobra.OnlyValidArgs()`

File: cli/cmd/completion.go (hunk 3 above)
Comment (siggy) [reply cli/cmd/completion.go]: will do. was reluctant to make this too verbose, but then realized:
```
$ kubectl completion -h|wc -l
      37
```

**Comment by siggy** [comment]
added example output:
```bash
$ conduit completion -h
Output completion code for the specified shell (bash or zsh).

Usage:
  conduit completion [bash|zsh] [flags]

Examples:
  # bash >= 4.0
  source <(conduit completion bash)

  # bash <= 3.2
  source /dev/stdin <<< "$(conduit completion bash)"

  # zsh
  source <(conduit completion zsh)

  # zsh on mac / oh-my-zsh
  conduit completion zsh > "${fpath[1]}/_conduit"

Flags:
  -h, --help   help for completion

Global Flags:
  -n, --conduit-namespace string   namespace in which Conduit is installed (default "conduit")
```

**Comment by pcalcado** [comment]
@siggy Works 💯  for oh-my-zsh:

```bash
pcalcado@arbeitsziege.local:/Users/pcalcado  $ conduit completion
Error: accepts 1 arg(s), received 0
Usage:
  conduit completion [bash|zsh] [flags]

Examples:
  # bash >= 4.0
  source <(conduit completion bash)

  # bash <= 3.2
  source /dev/stdin <<< "$(conduit completion bash)"

  # zsh
  source <(conduit completion zsh)

  # zsh on mac / oh-my-zsh
  conduit completion zsh > "${fpath[1]}/_conduit"

Flags:
  -h, --help   help for completion

Global Flags:
  -n, --conduit-namespace string   namespace in which Conduit is installed (default "conduit")

pcalcado@arbeitsziege.local:/Users/pcalcado  $ conduit completion zsh > "${fpath[1]}/_conduit"
pcalcado@arbeitsziege.local:/Users/pcalcado  $ zsh
pcalcado@arbeitsziege.local:/Users/pcalcado  $ conduit [tab] [tab]
completion  dashboard   get         help        inject      install     stat        tap         version
```


But I am still having troubles with bash:
### bash
```bash
pcalcado@arbeitsziege.local:/Users/pcalcado  $ conduit completion
Error: accepts 1 arg(s), received 0
Usage:
  conduit completion [bash|zsh] [flags]

Examples:
  # bash >= 4.0
  source <(conduit completion bash)

  # bash <= 3.2
  source /dev/stdin <<< "$(conduit completion bash)"

  # zsh
  source <(conduit completion zsh)

  # zsh on mac / oh-my-zsh
  conduit completion zsh > "${fpath[1]}/_conduit"

Flags:
  -h, --help   help for completion

Global Flags:
  -n, --conduit-namespace string   namespace in which Conduit is installed (default "conduit")

pcalcado@arbeitsziege.local:/Users/pcalcado  $ bash [tab] [tab]
bash-3.2$  source /dev/stdin <<< "$(conduit completion bash)"
bash-3.2$ conduit bash: _get_comp_words_by_ref: command not found
bash: _get_comp_words_by_ref: command not found

.CFUserTextEncoding          .gsutil/                     .spacemacs                   Applications/
```

While bash-3.2 is pretty old, it ships with Mac OS and it seems like Apple won't update it:
https://apple.stackexchange.com/a/197172/22068

But I found this on the kubectl completion, maybe we should follow suit?

https://github.com/kubernetes/kubernetes/blob/12d73c31a97188286e7f4c8a3aec7fcd38a76a83/pkg/kubectl/cmd/completion.go#L59

**Comment by siggy** [comment]
@pcalcado pushed better instructions...

```
$ conduit completion -h
Output completion code for the specified shell (bash or zsh).

Usage:
  conduit completion [bash|zsh] [flags]

Examples:
  # bash <= 3.2
  source /dev/stdin <<< "$(conduit completion bash)"

  # bash >= 4.0
  source <(conduit completion bash)

  # bash <= 3.2 on osx
  brew install bash-completion # ensure you have bash-completion 1.3+
  conduit completion bash > $(brew --prefix)/etc/bash_completion.d/conduit

  # bash >= 4.0 on osx
  brew install bash-completion@2
  conduit completion bash > $(brew --prefix)/etc/bash_completion.d/conduit

  # zsh
  source <(conduit completion zsh)

  # zsh on osx / oh-my-zsh
  conduit completion zsh > "${fpath[1]}/_conduit"

Flags:
  -h, --help   help for completion

Global Flags:
  -n, --conduit-namespace string   namespace in which Conduit is installed (default "conduit")
```

**Pull request merged**